zebra-network/src/peer/connection.rs

```diff
@@ -598,31 +602,38 @@ where
                 self.peer_tx
                     .send(Message::Tx(transaction))
                     .await
-                    .map_err(|e| e.into())
+                    .map_err(|e| (e.into(), None))
                     .map(|()| AwaitingRequest)
             }
             (AwaitingRequest, AdvertiseTransactions(hashes)) => {
                 let _ = tx.send(Ok(Response::Nil));
                 self.peer_tx
                     .send(Message::Inv(hashes.iter().map(|h| (*h).into()).collect()))
                     .await
-                    .map_err(|e| e.into())
+                    .map_err(|e| (e.into(), None))
                     .map(|()| AwaitingRequest)
             }
             (AwaitingRequest, AdvertiseBlock(hash)) => {
                 let _ = tx.send(Ok(Response::Nil));
                 self.peer_tx
                     .send(Message::Inv(vec![hash.into()]))
                     .await
-                    .map_err(|e| e.into())
+                    .map_err(|e| (e.into(), None))
                     .map(|()| AwaitingRequest)
             }
         } {
             Ok(new_state) => {
                 self.state = new_state;
                 self.request_timer = Some(sleep(constants::REQUEST_TIMEOUT));
             }
-            Err(e) => self.fail_with(e),
+            Err((e, Some(tx))) => {
+                // TODO: work out how to re-use `e`
+                // copy and clone don't work here, because some of the source
+                // errors don't implement copy or clone
+                tx.send(Err(SharedPeerError::from(e)));
+                self.fail_with(e);
```

Comment:
@yaahc I don't know how to use `e` in both `tx.send` and `self.fail_with`.

I can't work out how to impl `Copy` or `Clone` for `e`, because some of its sources don't impl `Copy` or `Clone`.

Reply:
I think the solution here is to convert to SharedPeerError, and change fail_with to take `Into<SharedPeerError>`.

Reply:
Sounds good. I'm pretty sure `SharedPeerError` only exists in the first place to make `e` clone, so just converting it a little earlier makes perfect sense to me.

Reply:
Fixed in 12fec92cc43e198127d567498809c215f96a57c4